core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go

```diff
@@ -141,26 +117,40 @@ func (hw *BlockSubscriber) Start(_ context.Context) error {
 	return hw.sync.StartOnce("BlockSubscriber", func() error {
 		hw.mu.Lock()
 		defer hw.mu.Unlock()
-		_, hw.unsubscribe = hw.hb.Subscribe(&headWrapper{headC: hw.headC})
 		hw.ctx, hw.cancel = context.WithCancel(context.Background())
 
+		// initialize the blocks map with the recent blockHistorySize blocks
+		blocks, err := hw.getBlockRange()
+		if err != nil {
+			hw.lggr.Errorf("failed to get block range", err)
+		}
+		err = hw.initializeBlocks(blocks)
+		if err != nil {
+			hw.lggr.Errorf("failed to get log poller blocks", err)
+		}
+
+		_, hw.unsubscribe = hw.hb.Subscribe(&headWrapper{headC: hw.headC})
+
 		// poll from head broadcaster channel and push to subscribers
 		{
 			go func(ctx context.Context) {
 				for {
 					select {
-					case bk := <-hw.headC:
+					case h := <-hw.headC:
 						hw.mu.Lock()
-						if hw.lastClearedBlock == 0 {
-							hw.lastClearedBlock = bk.block - 1
-							hw.lggr.Infof("lastClearedBlock is %d", hw.lastClearedBlock)
+						// head parent is a linked list with EVM finality depth
+						// when re-org happens, new heads will have pointers to the new blocks
+						for cp := h; cp != nil; cp = cp.Parent {
+							if cp != h && hw.blocks[cp.Number] != cp.Hash.Hex() {
+								hw.lggr.Warnf("overriding block %d old hash %s with new hash %s due to re-org", cp.Number, hw.blocks[cp.Number], cp.Hash.Hex())
+							}
+							hw.blocks[cp.Number] = cp.Hash.Hex()
 						}
-						hw.blocksFromBroadcaster[bk.block] = bk.hash
-						hw.lggr.Infof("blocksFromBroadcaster block %d hash is %s", bk.block, bk.hash.String())
+						hw.lggr.Infof("blocks block %d hash is %s", h.Number, h.Hash.Hex())
 
-						history := hw.buildHistory(bk.block)
+						history := hw.buildHistory(h.Number)
 
-						hw.lastSentBlock = bk.block
+						hw.lastSentBlock = h.Number
 						hw.lggr.Infof("lastSentBlock is %d", hw.lastSentBlock)
 						// send history to all subscribers
 						for _, subC := range hw.subscribers {
```

Comment:
Writing on channels might block, wrap with a select to ensure that we won't get affected by some subscriber:
```
select {
  case subC <- history:
  default:
    hw.lggr.Warnf("subscriber channel is full, dropping block history with length %d", len(history))
}
```

Reply:
done

Reply:
good idea